paasta_tools/autoscaling/autoscaling_cluster_lib.py

```diff
@@ -520,16 +520,26 @@ def downscale_aws_resource(self, filtered_slaves, current_capacity, target_capac
             current_capacity = new_capacity
             mesos_state = get_mesos_master().state_summary()
             if filtered_sorted_slaves:
+                good_ids = [s['id'] for s in mesos_state.get('slaves', [])]
+                # Filter out slaves no longer in mesos state
+                updated_filtered_slaves = [
+                    slave for slave in filtered_sorted_slaves
+                    if slave.instance_id in good_ids
+                ]
+                # Get updated task counts for slaves from state
                 task_counts = get_mesos_task_count_by_slave(
                     mesos_state,
                     slaves_list=[
                         {'task_counts': slave.task_counts}
-                        for slave in filtered_sorted_slaves
+                        for slave in updated_filtered_slaves
                     ],
                 )
-                for i, slave in enumerate(filtered_sorted_slaves):
+                # Update the task counts in the slaves list
+                for i, slave in enumerate(updated_filtered_slaves):
                     slave.task_counts = task_counts[i]['task_counts']
-            filtered_slaves = filtered_sorted_slaves
+            else:
+                updated_filtered_slaves = filtered_sorted_slaves
+            filtered_slaves = updated_filtered_slaves
```

Comment:
I don't really understand. What does this part of the code "do"?
It doesn't look like it is actually changing state, and the task_counts variables are not used later on?

Reply:
This part has confused me for a while, but I think I figured out what it does.  It basically gets a list of tasks from state_summary, and then updates the counts in the slave list, so next time through the loop when we call `ec2_fitness.sort_by_ec2_fitness`, it will correctly sort by the number of running tasks.  

This is useful here, since after stopping a slave its tasks may be moved elsewhere, and the fitness sort changes.  But now that I think of it it's almost useless on the async scaledown PR, since few or no slaves will actually be killed between iterations of the loop (in that branch).